### ISSUE TITLE
Error on reading BOM in snippets file

### ISSUE THREAD

**Issue by bergentroll** [SirVer/ultisnips#1365]
**Expected behavior:**
Using UTF-8 encoded snippets file with BOM.

**Actual behavior:**
Error on switching to insert mode.
```
Following is the full stack trace:
Traceback (most recent call last):
  File "/usr/share/vim/vimfiles/pythonx/UltiSnips/err_to_scratch_buffer.py", line 18, in wrapper
    return func(self, *args, **kwds)
  File "/usr/share/vim/vimfiles/pythonx/UltiSnips/snippet_manager.py", line 905, in _track_change
    self._try_expand(autotrigger_only=True)
  File "/usr/share/vim/vimfiles/pythonx/UltiSnips/snippet_manager.py", line 749, in _try_expand
    snippets = self._snips(before, False, autotrigger_only)
  File "/usr/share/vim/vimfiles/pythonx/UltiSnips/snippet_manager.py", line 629, in _snips
    source.ensure(filetypes)
  File "/usr/share/vim/vimfiles/pythonx/UltiSnips/snippet/source/file/base.py", line 31, in ensure
    self._load_snippets_for(ft)
  File "/usr/share/vim/vimfiles/pythonx/UltiSnips/snippet/source/file/base.py", line 53, in _load_snippets_for
    self._parse_snippets(ft, fn)
  File "/usr/share/vim/vimfiles/pythonx/UltiSnips/snippet/source/file/base.py", line 69, in _parse_snippets
    raise SnippetSyntaxError(filename, line_index, msg)
UltiSnips.snippet.source.file.base.SnippetSyntaxError: Invalid line '\ufeffsnippet sign "My signature including mailbox"' in ~/sources/dotfiles/.vim/UltiSnips/all.snippets:1
```

**Steps to reproduce**
1. Create UTF-8 encoded (`set encoding=utf-8`) snippets file (e.g. `all.snippets`).
2. Add BOM (`set bomb`).
3. Try to involve the snippets file.

-----
- **Operating System**: GNU/Linux
- **Vim Version**: 8.2
- **UltiSnips Version**: 3.2
- **Python inside Vim**: 3.9.5 (system)

**Comment by SirVer** [comment]
Can you explain why you are using a BOM on your snippet files? It should not be set on UTF files.

Wikipedia states: https://en.wikipedia.org/wiki/Byte_order_mark

> BOM use is optional. Its presence interferes with the use of UTF-8 by software that does not expect non-ASCII bytes at the start of a file but that could otherwise handle the text stream.

In general it seems possible that UltiSnips just ignores the first three bytes of a file if it looks like a BOM, but that is a workaround which I'd like to avoid. Hence my question what you plan to achieve with having the BOM there?

**Comment by bergentroll** [comment]
That is not "should not", it is just optional.

My `.vimrc` sets the BOM automatically for unicode files. Motivation: for some software it helps to determine a file encoding. E.g. I may send some UTF-8 text file with LF line endings to a colleague who uses an MS Windows machine and say to open it with Firefox browser. With the BOM browser will handle such file correct respecting national alphabet and line breaks.

The reason to notify you about the issue: majority of tools works well with the BOM including browsers, compilers, CPython.

It seems there is no need for workaround, it is possible just to do `s/utf-8/utf-8-sig/` here
https://github.com/SirVer/ultisnips/blob/b6c9a814a4fdfcad3ea53d8888746803a7218a9e/pythonx/UltiSnips/snippet/source/file/base.py#L61

It should work transparent for BOM-less files.
https://stackoverflow.com/a/44573867

Actually I am a bit surprised UltiSnips forces `encoding` while `open` a file. For me seems natural to allow Python to expect environment.

**Comment by SirVer** [comment]
> Actually I am a bit surprised UltiSnips forces encoding while open a file. For me seems natural to allow Python to expect environment.

that backfired in the past, that is why this encoding has been added. 

> It seems there is no need for workaround, it is possible just to do s/utf-8/utf-8-sig/ here

That sounds worth a try. Could you bring a patch? Otherwise I can do it, but currently I am a bit too busy for UltiSnips.

**Comment by bergentroll** [comment]
Sure! #1366

**Comment by SirVer** [comment]
Fixed in #1366
